File: objc/objcsearch/Tests/objcsearchTests/SearchResultTests.m
//
//  SearchResultTests.m
//  objcsearch_tests
//
//  Created by Cary Clark on 11/12/18.
//  Copyright © 2018 Cary Clark. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "common.h"
#import "FileResult.h"
#import "Regex.h"
#import "SearchResult.h"
#import "SearchResultFormatter.h"
#import "SearchSettings.h"

@interface SearchResultTests : XCTestCase

@end

@implementation SearchResultTests

- (void)setUp {
    [super setUp];
}

- (void)tearDown {
    // Put teardown code here. This method is called after the invocation of each test method in the class.
}

- (void)testSingleLineSearchResult {
    SearchSettings *settings = [[SearchSettings alloc] init];
    settings.colorize = false;
    SearchResultFormatter *formatter = [[SearchResultFormatter alloc] initWithSettings:settings];
    Regex *regex = [[Regex alloc] initWithPattern:@"Search"];
    FileResult *fr = [[FileResult alloc] initWithFilePath:@"~/src/xsearch/csharp/CsSearch/CsSearch/Searcher.cs" fileType:FileTypeCode fileSize:0 lastMod:nil];
    long lineNum = 10;
    long matchStartIndex = 15;
    long matchEndIndex = 23;
    NSString *line = @"\tpublic class Searcher\n";

    SearchResult *result = [[SearchResult alloc] initWithPattern:[regex pattern]
                                                            file:fr
                                                         lineNum:lineNum
                                                 matchStartIndex:matchStartIndex
                                                   matchEndIndex:matchEndIndex
                                                            line:line
                                                     linesBefore:[[NSArray alloc] init]
                                                      linesAfter:[[NSArray alloc] init]];

    NSString *trimmedLine = [line stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    NSString *expectedOutput = [NSString stringWithFormat:@"%@: %lu: [%lu:%lu]: %@", [fr description], lineNum,
                                matchStartIndex, matchEndIndex, trimmedLine];
    NSString *output = [formatter format:result];
    XCTAssert([output isEqualToString:expectedOutput]);
}

- (void)testBinaryFileSearchResult {
    SearchSettings *settings = [[SearchSettings alloc] init];
    SearchResultFormatter *formatter = [[SearchResultFormatter alloc] initWithSettings:settings];
    Regex *regex = [[Regex alloc] initWithPattern:@"Search"];
    FileResult *fr = [[FileResult alloc] initWithFilePath:@"~/src/xsearch/csharp/CsSearch/CsSearch/Searcher.exe" fileType:FileTypeBinary fileSize:0 lastMod:nil];
    long lineNum = 0;
    long matchStartIndex = 0;
    long matchEndIndex = 0;
    
    SearchResult *result = [[SearchResult alloc] initWithPattern:[regex pattern]
                                                            file:fr
                                                         lineNum:lineNum
                                                 matchStartIndex:matchStartIndex
                                                   matchEndIndex:matchEndIndex
                                                            line:NULL
                                                     linesBefore:[[NSArray alloc] init]
                                                      linesAfter:[[NSArray alloc] init]];

    NSString *expectedOutput = [NSString stringWithFormat:@"%@ matches at [%lu:%lu]", [fr description],
                                matchStartIndex, matchEndIndex];
    NSString *output = [formatter format:result];
    XCTAssert([output isEqualToString:expectedOutput]);
}

- (void)testMultiLineSearchResult {
    SearchSettings *settings = [[SearchSettings alloc] init];
    SearchResultFormatter *formatter = [[SearchResultFormatter alloc] initWithSettings:settings];
    Regex *regex = [[Regex alloc] initWithPattern:@"Search"];
    NSString *filePath = @"~/src/xsearch/csharp/CsSearch/CsSearch/Searcher.cs";
    FileResult *fr = [[FileResult alloc] initWithFilePath:filePath fileType:FileTypeText fileSize:0 lastMod:nil];
    long lineNum = 10;
    long matchStartIndex = 15;
    long matchEndIndex = 23;
    NSString *line = @"\tpublic class Searcher\n";
    NSArray<NSString*> *linesBefore = @[@"namespace CsSearch\n", @"{\n"];
    NSArray<NSString*> *linesAfter = @[@"\t{\n", @"\t\tprivate readonly FileTypes _fileTypes;\n"];

    SearchResult *result = [[SearchResult alloc] initWithPattern:[regex pattern]
                                                            file:fr
                                                         lineNum:lineNum
                                                 matchStartIndex:matchStartIndex
                                                   matchEndIndex:matchEndIndex
                                                            line:line
                                                     linesBefore:linesBefore
                                                      linesAfter:linesAfter];
    
    NSString *expectedOutput = [NSString stringWithFormat:@"================================================================================\n"
    "%@: %lu: [%lu:%lu]\n"
    "--------------------------------------------------------------------------------\n"
    "   8 | namespace CsSearch\n"
    "   9 | {\n"
    "> 10 | \tpublic class Searcher\n"
    "  11 | \t{\n"
    "  12 | \t\tprivate readonly FileTypes _fileTypes;\n", filePath, lineNum, matchStartIndex, matchEndIndex];

    NSString *output = [formatter format:result];
    logMsg(@"\noutput:");
    logMsg(output);
    logMsg(@"\nexpectedOutput:");
    logMsg(expectedOutput);
    XCTAssert([output isEqualToString:expectedOutput]);
}

@end

File: swift/swiftsearch/Tests/swiftsearchTests/SearchResultTests.swift
//
//  SearchResultTests.swift
//  swiftsearch
//
//  Created by Cary Clark on 5/18/15.
//  Copyright (c) 2015 Cary Clark. All rights reserved.
//

import Cocoa
import XCTest

import swiftsearch
import swiftfind

class SearchResultTests: XCTestCase {
    func testSingleLineSearchResult() {
        let pattern = "Search"
        let filepath = "~/src/xsearch/csharp/CsSearch/CsSearch/Searcher.cs"
        let fileResult = FileResult(filePath: filepath, fileType: FileType.code, fileSize: 0)
        let lineNum = 10
        let matchStartIndex = 15
        let matchEndIndex = 21
        let line = "\tpublic class Searcher\n"
        let searchResult = SearchResult(
            searchPattern: pattern,
            file: fileResult,
            lineNum: lineNum,
            matchStartIndex: matchStartIndex,
            matchEndIndex: matchEndIndex,
            line: line,
            linesBefore: [],
            linesAfter: []
        )
        let expectedLine = line.trimmingCharacters(in: whitespace as CharacterSet)
        let expectedOutput = "\(filepath): \(lineNum): [\(matchStartIndex):\(matchEndIndex)]: \(expectedLine)"

        let settings = SearchSettings()
        settings.colorize = false
        let formatter = SearchResultFormatter(settings: settings)

        let output = formatter.format(result: searchResult)
        XCTAssertEqual(expectedOutput, output, "single-line searchResult matches expected")
    }

    func testSingleLineSearchResultLongerThanMaxLineLength() {
        let pattern = "maxlen"
        let file = "./maxlen.txt"
        let fileResult = FileResult(filePath: file, fileType: FileType.text, fileSize: 0)
        let lineNum = 1
        let matchStartIndex = 53
        let matchEndIndex = 59
        let line = "0123456789012345678901234567890123456789012345678901maxlen8901234567890123456789012345678901234567890123456789"
        let linesBeforeAfter: [String] = []
        let searchResult = SearchResult(
            searchPattern: pattern,
            file: fileResult,
            lineNum: lineNum,
            matchStartIndex: matchStartIndex,
            matchEndIndex: matchEndIndex,
            line: line,
            linesBefore: linesBeforeAfter,
            linesAfter: linesBeforeAfter
        )
        let expectedLine = "...89012345678901234567890123456789012345678901maxlen89012345678901234567890123456789012345678901..."
        let expectedOutput = "\(file): \(lineNum): [\(matchStartIndex):\(matchEndIndex)]: \(expectedLine)"

        let settings = SearchSettings()
        settings.colorize = false
        settings.maxLineLength = 100
        let formatter = SearchResultFormatter(settings: settings)

        let output = formatter.format(result: searchResult)
        XCTAssertEqual(expectedOutput, output, "single-line longer than maxLineLength searchResult matches expected")
    }

    func testSingleLineSearchResultLongerColorize() {
        let pattern = "maxlen"
        let file = "./maxlen.txt"
        let fileResult = FileResult(filePath: file, fileType: FileType.text, fileSize: 0)
        let lineNum = 1
        let matchStartIndex = 53
        let matchEndIndex = 59
        let line = "0123456789012345678901234567890123456789012345678901maxlen8901234567890123456789012345678901234567890123456789"
        let linesBeforeAfter: [String] = []
        let searchResult = SearchResult(
            searchPattern: pattern,
            file: fileResult,
            lineNum: lineNum,
            matchStartIndex: matchStartIndex,
            matchEndIndex: matchEndIndex,
            line: line,
            linesBefore: linesBeforeAfter,
            linesAfter: linesBeforeAfter
        )
        let expectedLine = "...89012345678901234567890123456789012345678901" +
            ConsoleColor.GREEN +
            "maxlen" +
            ConsoleColor.RESET +
            "89012345678901234567890123456789012345678901..."
        let expectedOutput = "\(file): \(lineNum): [\(matchStartIndex):\(matchEndIndex)]: \(expectedLine)"

        let settings = SearchSettings()
        settings.colorize = true
        settings.maxLineLength = 100
        let formatter = SearchResultFormatter(settings: settings)

        let output = formatter.format(result: searchResult)
        XCTAssertEqual(expectedOutput, output, "single-line colorized searchResult matches expected")
    }

    func testSearchResultMatchLongerColorize() {
        let pattern = "\\d+maxlen\\d+"
        let file = "./maxlen.txt"
        let fileResult = FileResult(filePath: file, fileType: FileType.text, fileSize: 0)
        let lineNum = 1
        let matchStartIndex = 1
        let matchEndIndex = 110
        let line = "0123456789012345678901234567890123456789012345678901maxlen8901234567890123456789012345678901234567890123456789"
        let linesBeforeAfter: [String] = []
        let searchResult = SearchResult(
            searchPattern: pattern,
            file: fileResult,
            lineNum: lineNum,
            matchStartIndex: matchStartIndex,
            matchEndIndex: matchEndIndex,
            line: line,
            linesBefore: linesBeforeAfter,
            linesAfter: linesBeforeAfter
        )
        let expectedLine = ConsoleColor.GREEN +
            "0123456789012345678901234567890123456789012345678901maxlen890123456789012345678901234567890123456" +
            ConsoleColor.RESET +
            "..."
        let expectedOutput = "\(file): \(lineNum): [\(matchStartIndex):\(matchEndIndex)]: \(expectedLine)"

        let settings = SearchSettings()
        settings.colorize = true
        settings.maxLineLength = 100
        let formatter = SearchResultFormatter(settings: settings)

        let output = formatter.format(result: searchResult)
        XCTAssertEqual(expectedOutput, output, "colorized match longer than maxLineLength searchResult matches expected")
    }

    func testSearchResult2MatchLongerColorize() {
        let pattern = "\\d+maxlen\\d+"
        let file = "./maxlen.txt"
        let fileResult = FileResult(filePath: file, fileType: FileType.text, fileSize: 0)
        let lineNum = 1
        let matchStartIndex = 11
        let matchEndIndex = 120
        let line = "ABCDEFGHIJ0123456789012345678901234567890123456789012345678901maxlen8901234567890123456789012345678901234567890123456789ABCDEFGHIJ"
        let linesBeforeAfter: [String] = []
        let searchResult = SearchResult(
            searchPattern: pattern,
            file: fileResult,
            lineNum: lineNum,
            matchStartIndex: matchStartIndex,
            matchEndIndex: matchEndIndex,
            line: line,
            linesBefore: linesBeforeAfter,
            linesAfter: linesBeforeAfter
        )
        let expectedLine = "..." +
            ConsoleColor.GREEN +
            "3456789012345678901234567890123456789012345678901maxlen890123456789012345678901234567890123456" +
            ConsoleColor.RESET +
            "..."
        let expectedOutput = "\(file): \(lineNum): [\(matchStartIndex):\(matchEndIndex)]: \(expectedLine)"

        let settings = SearchSettings()
        settings.colorize = true
        settings.maxLineLength = 100
        let formatter = SearchResultFormatter(settings: settings)

        let output = formatter.format(result: searchResult)
        XCTAssertEqual(expectedOutput, output, "colorized match 2 longer than maxLineLength searchResult matches expected")
    }

    func testBinaryFileSearchResult() {
        let pattern = "Search"
        let file = "~/src/xsearch/csharp/CsSearch/CsSearch/Searcher.exe"
        let fileResult = FileResult(filePath: file, fileType: FileType.binary, fileSize: 0)
        let lineNum = 0
        let matchStartIndex = 0
        let matchEndIndex = 0
        let searchResult = SearchResult(
            searchPattern: pattern,
            file: fileResult,
            lineNum: lineNum,
            matchStartIndex: matchStartIndex,
            matchEndIndex: matchEndIndex,
            line: "",
            linesBefore: [],
            linesAfter: []
        )
        let expectedOutput = "\(file) matches at [0:0]"

        let settings = SearchSettings()
        let formatter = SearchResultFormatter(settings: settings)

        let output = formatter.format(result: searchResult)
        XCTAssertEqual(expectedOutput, output, "binary searchResult matches expected")
    }

    func testMultiLineSearchResult() {
        let pattern = "Search"
        let file = "~/src/xsearch/csharp/CsSearch/CsSearch/Searcher.cs"
        let fileResult = FileResult(filePath: file, fileType: FileType.text, fileSize: 0)
        let lineNum = 10
        let matchStartIndex = 15
        let matchEndIndex = 23
        let line = "\tpublic class Searcher"
        let linesBefore = ["namespace CsSearch", "{"]
        let linesAfter = ["\t{", "\t\tprivate readonly FileTypes _fileTypes;"]
        let searchResult = SearchResult(
            searchPattern: pattern,
            file: fileResult,
            lineNum: lineNum,
            matchStartIndex: matchStartIndex,
            matchEndIndex: matchEndIndex,
            line: line,
            linesBefore: linesBefore,
            linesAfter: linesAfter
        )
        let expectedOutput =
            """
            ================================================================================
            \(file): \(lineNum): [\(matchStartIndex):\(matchEndIndex)]
            --------------------------------------------------------------------------------
               8 | namespace CsSearch
               9 | {
            > 10 | 	public class Searcher
              11 | 	{
              12 | 		private readonly FileTypes _fileTypes;

            """

        let settings = SearchSettings()
        settings.colorize = false
        let formatter = SearchResultFormatter(settings: settings)

        let output = formatter.format(result: searchResult)
        XCTAssertEqual(expectedOutput, output, "multiline searchResult matches expected")
    }
}
